src/amber/cli/templates/auth/src/controllers/{{name}}_controller.cr.ecr

```diff
@@ -0,0 +1,45 @@
+class <%= class_name %>Controller < ApplicationController
+  def show
+    if (<%= @name %> = context.current_<%= @name %>)
+      render("show.slang")
+    else
+      flash[:info] = "Must be logged in"
+      redirect_to "/signin"
+    end
+  end
+
+  def edit
+    if (<%= @name %> = context.current_<%= @name %>)
+      render("edit.slang")
+    else
+      flash[:info] = "Must be logged in"
+      redirect_to "/signin"
+    end
+  end
+
+  def update
+    if (<%= @name %> = context.current_<%= @name %>)
+      update_<%= @name %>(<%= @name %>)
+    else
+      flash[:info] = "Must be logged in"
+      redirect_to "/signin"
+    end
+  end
+
+  private def update_<%= @name %>(<%= @name %>)
+    <%= @name %>.set_attributes(<%= @name %>_params.to_h)
+    if <%= @name %>.valid? && <%= @name %>.save
+      flash[:success] = "Updated Profile successfully."
+      redirect_to "/profile"
+    else
+      flash[:danger] = "Could not update Profile!"
+      render("edit.slang")
+    end
+  end
+
+  private def <%= @name %>_params
+    params.validation do
+      required(:email) {|f| !f.nil? && !f.empty? }
+    end
+  end
```

Comment:
This probably needs to be revisited. I noticed that if I called `user_params.validate!` it gave a runtime exception, so I need to know more about how these should be written and used.

Reply:
I don't know how this works as well.  @eliasjpr ?

Reply:
It looks like calling `params.valid?` in combination with `params.to_h` may be the way to go. @eliasjpr can you comment?

Reply:
In a nutshell this is how it works.

```crystal
# This defines a validation to perform
params.validation do
  required(:email, "Custom error message here") {|f| !f.nil? && !f.empty? }
end

# Returns boolean
params.valid?     

# Success: returns a hash with only the fields defined in the validation
# Lets assume token, password and email exists in the request params 
# but only email is returned since password and token are not defined in the validation
params.validate! # => { :email => "some@email.com" } 

# Failure: returns exception and `params.errors` will have the failed fields
params.validate! # => `params.errors => [ { :email => "Custom error message here"}]`
```

Reply:
@eliasjpr - Can you take a look at this file: https://github.com/marksiemers/auth-test/blob/7cc94319054a7a1971c4d269cbf6bacf1cc6759c/src/controllers/user_controller.cr

Overall, does this look good? The use of `before_all`, the way params validations work, the overall design of the controller and actions.

Reply:
After tracing the `user` argument it seems that it would always be present by the time is performing the update. 

This how I might approach it personally. Maybe you can get an idea or two from it.
```crystal
class UserController < ApplicationController
  # Since the context already exists you can assign the context to the user in the property
  property current_user : User? = context.current_user

  before_action do
    all { user_signed_in? }
  end

  def show
    render("show.slang") if (user = @current_user)
  end

  def edit
    render("edit.slang") if (user = @current_user)
  end

  def update
    if update_user?
      flash[:success] = "Updated Profile successfully."
      redirect_to "/profile"
    else
      flash[:danger] = "Could not update Profile!"
      render("edit.slang")
    end
  end

  private def update_user(user)
    current_user.set_attributes(user_params.validate!)
    # Granite::ORM It would be nice to be able to do something like current_user.save!  where `save!` runs the validations.
    current_user.valid? && current_user.save
  end

  private def user_params
    params.validation do
      required(:email) {|f| !f.nil? && !f.empty? }
      optional(:password) {|f| !f.nil? && !f.empty? }
    end
  end
  
  # Previously this method was had two responsability assign the current user and redirect
  # Now it just checks if the current user is signed in.
  private def user_signed_in?
    unless current_user
      flash[:info] = "Must be logged in"
      redirect_to "/signin"
    end
  end
end
```

Reply:
> After tracing the user argument it seems that it would always be present by the time is performing the update.

This is true, I don't know if the compiler can track it though. I can put this syntax in and see what happens. 

The same is true for `edit` and `show` (@current_user will always be something), but with the current setup, the compiler can't track that.

Reply:
It also seems that people could benefit from having a user_signed_in? and current_user in application controller. Since people would exercise these methods is other places as well.

Reply:
This two methods should definitely be added to application_controller.cr
```crystal
def current_user
  @current_user ||= context.current_user
end

private def redirect_signed_out_user 
    unless current_user
      flash[:info] = "Must be logged in"
      redirect_to "/signin"
    end
end
```